pennylane/math/ode.py

```diff
@@ -0,0 +1,163 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# TODO: This code is heavily inspired by https://github.com/google/jax/blob/main/jax/experimental/ode.py
+# need to make sure copyright is not infringed
+"""Fix step size ODE solver"""
+
+import warnings
+from functools import partial
+
+import jax
+import jax.numpy as jnp
+from jax import linear_util as lu
+from jax.flatten_util import ravel_pytree
+
+
+@partial(jax.jit, static_argnums=0)
+def odeint(func, y0, ts, *args):
+    """Fix step size ODE solver
+
+    Solves the initial value problem (IVP) of an ordinary differential equation (ODE)
+
+    .. math:: \frac{dy}{dt} = f(y, t), y(t_0) = y_0
+
+    using an `explicit Runge-Kutta method <https://en.wikipedia.org/wiki/Runge%E2%80%93Kutta_methods>`_.
+
+    Args:
+        func (callable): ``f(y, t, *args)`` defining the ODE
+        y0 (tensor_like): initial value ``y(t0) = y0)``
+        ts (tensor_like): finite time steps for the ODE solver to take.
+
+    .. note::
+
+        Unlike many standard implementations, this ``odeint`` solver does not adaptively choose the step
+        sizes, but rather a fix list of times for evaluation have to be provided in ``ts``.
+
+    **Example**
+
+    We can solve the time-dependent Schrodinger equation
+
+    .. math:: \frac{d}{dt}U = -i H(t) U
+
+    for a time-dependent Hamiltonian :math:`H(t) = X_0 X_1 + v \sin(t) Z_0 Y_1`
+    for the time window ``(t0, t1) = (0, 4)`` using ``odeint`` in the following way:
+
+    .. code:: python3
+
+        XX = qml.matrix(qml.PauliX(0) @ qml.PauliX(1))
+        ZY = qml.matrix(qml.PauliZ(0) @ qml.PauliY(1))
+
+        ts = jnp.linspace(0., 4., 20)
+        y0 = jnp.eye(2**2, dtype=complex)
+        params = jnp.ones(1, dtype=complex)
+
+        def func(y, t, params):
+            H = jnp.array([XX + params[0] * jnp.sin(t) * ZY])
+            return -1j * jnp.sum(H, axis=0) @ y
+
+        U = qml.math.odeint(func, y0, ts, params)
+    """
+    y0, unravel = ravel_pytree(y0)
+    func = ravel_first_arg(func, unravel)
+    out = _odeint(func, y0, ts, *args)
+    return unravel(out)
+
+def abs2(x):
+    if jnp.iscomplexobj(x):
+        return x.real ** 2 + x.imag ** 2
+    return x ** 2
+
+def _odeint(func, y0, ts, *args, atol=1e-8, rtol=1e-8):
+    def func_(y, t):
+        return func(y, t, *args)
+
+    def scan_fun(carry, t1):
+        y0, f0, t0 = carry
+        dt = t1 - t0
+        # not using y1_error and k atm
+        y1, f1, y1_error = runge_kutta_step(func_, y0, f0, t0, dt)
+
+        # check error
+        #def mean_error_ratio(error_estimate, rtol, atol, y0, y1):
+        err_tol = atol + rtol * jnp.maximum(jnp.abs(y0), jnp.abs(y1))
+        err_ratio = y1_error / err_tol.astype(y1_error.dtype)
+        mean_err_ratio = jnp.sqrt(jnp.mean(abs2(err_ratio)))
+        if mean_err_ratio > 1.:
```

Comment:
This fails unsurprisingly with 
```pycon
File ~/anaconda3/envs/pennylane/lib/python3.8/site-packages/jax/core.py:1264, in concretization_function_error.<locals>.error(self, arg)
   1263 def error(self, arg):
-> 1264   raise ConcretizationTypeError(arg, fname_context)

ConcretizationTypeError: Abstract tracer value encountered where concrete value is expected: Traced<ShapedArray(bool[])>with<DynamicJaxprTrace(level=1/1)>
The problem arose with the `bool` function. 
The error occurred while tracing the function scan_fun at /home/qottmann/Qottmann/Xanadu/pennylane/pennylane/math/ode.py:86 for scan. This concrete value was not available in Python because it depends on the values of the arguments 'carry' and 't1'.

See https://jax.readthedocs.io/en/latest/errors.html#jax.errors.ConcretizationTypeError
```

Is there a way to raise a warning like this in jax? pinging @PhilipVinc @rmoyard @dwierichs

Reply:
Is there the option to use `jax.lax.cond`, or is that overkill?

Reply:
Yep, something like that!

Reply:
Works in principle, but is there a way to obtain the values of traced arrays?
```pycon
UserWarning: The targeted error tolerance of atol = Traced<ShapedArray(float32[], weak_type=True)>with<DynamicJaxprTrace(level=2/1)> and rtol = Traced<ShapedArray(float32[], weak_type=True)>with<DynamicJaxprTrace(level=2/1)>is not reached with a relative error of Traced<ShapedArray(float32[1])>with<DynamicJaxprTrace(level=2/1)>
  warnings.warn(
```

Reply:
Or maybe add a check that the arrays are not abstract (which I guess only happens in the first execution for the jitting? Or am I making a conceptual mistake here?)

Is there a way to check if an array is abstract? I only find the recommendation to check `isinstance(array, jnp.ndarray)` but `jnp.ndarray` now includes tracer arrays and also it throws me an error https://github.com/google/jax/issues/2115

Reply:
Maybe somebody on the jax discussions has an answer / best practice: https://github.com/google/jax/discussions/13984

Reply:
@Qottmann callbacks and if/else statements are extremely slow in Jax, so I would suggest not using them here.

In some similar code I've written, I set a flag value with an integer (0,1,2...) depending on the error code (too large error, too small dt..) and every 50 iterations or so we check that flag if it's non zero we stop and throw the correct error

https://github.com/netket/netket/blob/aa7dd3701240f09868f0bf0880966c72fded1092/netket/experimental/dynamics/_rk_solver_structures.py#L230

Reply:
It seems that the `id_tap` callback function is quite efficient, given that it doesn't have to wait for a response.

We should however be very careful when using functions from inside `jax.experimental`, given that these will likely change in the future, adding breaking changes that might break pennylane.

Reply:
That is exactly the reason why we implement the ODE solver ourselves here. Are you referring to using `jax.experimental` in the tests?

Reply:
On topic of efficiency, I am currently doing some benchmarks testing how much of an impact the warning line has. I could see this not having a too dramatic impact since the passed values are only five scalars. The flag approach that Filippo is proposing seems also viable, though I have some uncertainties about implementation (e.g. how do you efficiently only check every X iterations in a `jax.lax.scan`?)

Reply:
Some preliminary benchmark for different qubit sizes, the effect of the warning seems minor:
![image](https://user-images.githubusercontent.com/43949391/212671627-99062ccc-e2da-47a6-8cf8-18a0a3e03672.png)

(more worrying is the difference to the jax implementation which should be slower if anything, but that is another story)

Reply:
I mark the question of performance issues due to id_tap as resolved now. Though the question why this simple odeint is performing slower than the jax odeint is still a mystery to me.